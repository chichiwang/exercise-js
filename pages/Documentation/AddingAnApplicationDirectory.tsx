import React from 'react';

import HeadingBlock from 'app/components/HeadingBlock';
import ExternalLink from 'app/components/ExternalLink';
import CopyButton, { ColorModes } from 'app/components/CopyButton';

import styles from './styles.module.scss';

function AddingAnApplicationDirectory() {
  return (
    <HeadingBlock
      level="3"
      id="Adding-an-Application-Directory"
      heading="Adding an Application Directory"
    >
      <p>
        To add additional application directories, or to remove existing ones:
      </p>

      <ul className={styles.list}>
        <li>Path aliases need to be updated</li>
        <li>Jest projects need to be updated</li>
      </ul>

      <p>
        {'The '}
        <span className={styles.highlight}>app/</span>
        {' and '}
        <span className={styles.highlight}>pages/</span>
        {` directories are set up with path aliases so that a module can be imported with
         absolute pathing rather than relative pathing:`}
      </p>

      <div className={styles.codeBlock}>
        <CopyButton
          className={styles.copyButton}
          textToCopy={'import Module from app/Module;\nimport PageModule from page/PageModule;'}
          colorMode={ColorModes.light}
        />
        <p>
          <span className={styles.red}>import</span>
          {' Module '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;app/module&apos;</span>
          ;
          <br />
          <span className={styles.red}>import</span>
          {' PageModule '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;page/PageModule&apos;</span>
          ;
        </p>
      </div>

      <p>
        If any directory is removed, or if a new directory needs an alias:
      </p>

      <ul className={styles.list}>
        <li>
          {'An '}
          <span className={styles.highlight}>alias</span>
          {' should be updated in the '}
          <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/webpack/resolve.js">
            Webpack configs
          </ExternalLink>
          .
        </li>
        <li>
          {'The '}
          <ExternalLink href="https://github.com/johvin/eslint-import-resolver-alias?tab=readme-ov-file#usage">
            alias mapping
          </ExternalLink>
          {' should be added to the '}
          <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/.eslintrc.js">
            eslint configuration
          </ExternalLink>
          {' to avoid linting errors when using the alias.'}
        </li>
        <li>
          {'The alias should be updated in the '}
          <ExternalLink href="https://www.typescriptlang.org/tsconfig#paths">
            paths configuration
          </ExternalLink>
          {' for TypeScript to avoid type errors.'}
        </li>
        <li>
          {'The alias should be updated in the '}
          <ExternalLink href="https://jestjs.io/docs/configuration#modulenamemapper-objectstring-string--arraystring">
            moduleNameMapper configuration
          </ExternalLink>
          {' in the '}
          <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/jest.config.js">
            Jest configuration
          </ExternalLink>
          {' to ensure module mocking will work in tests, and so there are no errors in using the alias.'}
          <ul className={styles.list}>
            <li>
              {'The directory should be updated in the '}
              <span className={styles.highlight}>appConfig.testMatch</span>
              {' array to ensure that the test runner knows which directories to cover.'}
            </li>
          </ul>
        </li>
      </ul>
    </HeadingBlock>
  );
}

export default AddingAnApplicationDirectory;
