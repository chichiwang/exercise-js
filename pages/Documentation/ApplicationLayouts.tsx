import React from 'react';

import HeadingBlock from 'app/components/HeadingBlock';
import ExternalLink from 'app/components/ExternalLink';
import CopyButton, { ColorModes } from 'app/components/CopyButton';

import styles from './styles.module.scss';

const dataRoutesSnippet = `// dataRoutes.js
import Layout from 'app/Layout';

import Home from 'pages/Home';
import AnotherPage from 'pages/AnotherPage';

const routes = [{
  Component: Layout,
  children: [{
    path: '/',
    Component: Home,
  }, {
    path: 'another-page',
    Component: AnotherPage,
  }],
}];`;

function ApplicationLayouts() {
  return (
    <HeadingBlock
      level="3"
      id="Application-Layouts"
      heading="Application Layouts"
    >
      <p>
        {'Layouts can be directly applied to routes in '}
        <ExternalLink href="https://reactrouter.com/en/main">React Router</ExternalLink>
        {' using '}
        <ExternalLink href="https://reactrouter.com/en/main/start/concepts#layout-routes">
          Layout Routes
        </ExternalLink>
        .
      </p>

      <div className={styles.codeBlock}>
        <CopyButton
          className={styles.copyButton}
          textToCopy={dataRoutesSnippet}
          colorMode={ColorModes.light}
        />
        <p>
          <span className={styles.grey}>&#47;&#47; dataRoutes.js</span>
          <br />
          <span className={styles.red}>import</span>
          {' '}
          <span className={styles.orange}>Layout</span>
          {' '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;app/Layout&apos;</span>
          ;
          <br />
          <br />
          <span className={styles.red}>import</span>
          {' '}
          <span className={styles.orange}>Home</span>
          {' '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;pages/Home&apos;</span>
          ;
          <br />
          <span className={styles.red}>import</span>
          {' '}
          <span className={styles.orange}>AnotherPage</span>
          {' '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;pages/AnotherPage&apos;</span>
          ;
          <br />
          <br />
          <span className={styles.red}>const</span>
          {' routes '}
          <span className={styles.blue}>=</span>
          {' [{'}
          <br />
          &nbsp;&nbsp;
          <span className={styles.darkBlue}>Component</span>
          {': '}
          <span className={styles.orange}>Layout</span>
          ,
          <br />
          &nbsp;&nbsp;
          <span className={styles.darkBlue}>children</span>
          {': [{'}
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>path</span>
          {': '}
          <span className={styles.blue}>&apos;/&apos;</span>
          ,
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>Component</span>
          {': '}
          <span className={styles.orange}>Home</span>
          ,
          <br />
          &nbsp;&nbsp;
          {'}, {'}
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>path</span>
          {': '}
          <span className={styles.blue}>&apos;/another-page&apos;</span>
          ,
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>Component</span>
          {': '}
          <span className={styles.orange}>AnotherPage</span>
          ,
          <br />
          &nbsp;&nbsp;
          {'}],'}
          <br />
          {'}];'}
        </p>
      </div>

      <p>
        {'In the boilerplate, a default Layout has been created in '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/Layout">
          app/Layout
        </ExternalLink>
        .
      </p>

      <p>
        {`One issue that has arisen from the usage of a layout route is that client-side navigation
         will not reset the browser scroll upon page navigation: clicking a link to another page will
         land the user on the new page, scrolled the same amount as on the previous page.`}
      </p>

      <p>
        {'A solution has been implemented in this boilerplate in the form of a '}
        <ExternalLink href="https://react.dev/learn/reusing-logic-with-custom-hooks">
          custom react hook
        </ExternalLink>
        {': '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/hooks/README.md#useresetscroll">
          useResetScroll
        </ExternalLink>
        {'. This hook needs to be invoked by any layouts in use, and '}
        <ExternalLink href="https://reactrouter.com/en/main/components/link">
          Links
        </ExternalLink>
        {' to pages that need the scroll reset need to pass a '}
        <ExternalLink href="https://reactrouter.com/en/main/components/link#state">
          state
        </ExternalLink>
        {' object with a property '}
        <span className={styles.highlight}>resetScroll</span>
        {' set to '}
        <span className={styles.highlight}>true</span>
        {'. A reusable component '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/components/InternalLink">
          InternalLink
        </ExternalLink>
        {' has been provided to automatically pass this state object property.'}
      </p>
    </HeadingBlock>
  );
}

export default ApplicationLayouts;
