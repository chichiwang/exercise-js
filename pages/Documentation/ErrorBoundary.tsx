import React from 'react';

import HeadingBlock from 'app/components/HeadingBlock';
import ExternalLink from 'app/components/ExternalLink';
import CopyButton, { ColorModes } from 'app/components/CopyButton';

import styles from './styles.module.scss';

const errorBoundarySnippet = `// app/dataRoutes/index.ts
import PageComponent from 'pages/pageComponent';

import withErrorBoundary from './withErrorBoundary';

const dataRoutes = [
  withErrorBoundary({
    path: '/path-to-page',
    Component: PageComponent,
  }),
];`;

function ErrorBoundary() {
  return (
    <HeadingBlock
      level="3"
      id="Error-Boundary"
      heading="Error Boundary"
    >
      <p>
        <b>Tamsui</b>
        {' implements a basic '}
        <ExternalLink href="https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary">
          React Error Boundary
        </ExternalLink>
        {' in '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/ErrorBoundary">
          app/ErrorBoundary
        </ExternalLink>
        {'. This error boundary is configured as an '}
        <ExternalLink href="https://reactrouter.com/en/main/route/error-element">
          errorElement
        </ExternalLink>
        {' in the '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/dataRoutes">
          dataRoutes
        </ExternalLink>
        {' object. It is recommended all root routes are wrapped in this error boundary. A utility, '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/dataRoutes/withErrorBoundary.tsx">
          withErrorBoundary
        </ExternalLink>
        {' is provided to more easily (and declaratively) extend routes with this '}
        <span className={styles.highlight}>errorElement</span>
        {' property:'}
      </p>

      <div className={styles.codeBlock}>
        <CopyButton
          className={styles.copyButton}
          textToCopy={errorBoundarySnippet}
          colorMode={ColorModes.light}
        />
        <p>
          <span className={styles.grey}>&#47;&#47; app/dataRoutes/index.ts</span>
          <br />
          <span className={styles.red}>import</span>
          {' '}
          <span className={styles.orange}>PageComponent</span>
          {' '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;pages/PageComponent&apos;</span>
          ;
          <br />
          <br />
          <span className={styles.red}>import</span>
          {' withErrorBoundary '}
          <span className={styles.red}>from</span>
          {' '}
          <span className={styles.blue}>&apos;./withErrorBoundary&apos;</span>
          ;
          <br />
          <br />
          <span className={styles.red}>const</span>
          {' dataRoutes '}
          <span className={styles.blue}>=</span>
          {' ['}
          <br />
          &nbsp;&nbsp;
          <span className={styles.purple}>withErrorBoundary</span>
          (&#123;
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>path</span>
          {': '}
          <span className={styles.blue}>&apos;/path-to-page&apos;</span>
          ,
          <br />
          &nbsp;&nbsp;&nbsp;&nbsp;
          <span className={styles.darkBlue}>Component</span>
          {': '}
          <span className={styles.orange}>PageComponent</span>
          ,
          <br />
          &nbsp;&nbsp;&#125;),
          <br />
          ];
        </p>
      </div>

      <p>
        {`On the server, if an error in rendering the application occurs it will
         redirect the request to `}
        <span className={styles.highlight}>/error</span>
        {', '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/server/appHandler.tsx">
          configured by default
        </ExternalLink>
        {' to be the '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/pages/ErrorPage">
          application&apos;s error page
        </ExternalLink>
        .
      </p>

      <p>
        {'On the client, an error in a page will render the '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/pages/ErrorPage">
          ErrorPage component
        </ExternalLink>
        {' as fallback.'}
      </p>

      <p>
        <b>To change the path to the error page</b>
        {': change the value of the constant '}
        <span className={styles.highlight}>errorPagePath</span>
        {' in '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/server/appHandler.tsx">
          server/appHandler.tsx
        </ExternalLink>
        {'. Make sure you define this route in the '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/app/dataRoutes/index.ts">
          app/dataRoutes/index.ts
        </ExternalLink>
        {' file.'}
      </p>

      <p>
        {'Alternatively, keep the error page path and just replace the '}
        <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/pages/ErrorPage">
          ErrorPage component
        </ExternalLink>
        .
      </p>
    </HeadingBlock>
  );
}

export default ErrorBoundary;
