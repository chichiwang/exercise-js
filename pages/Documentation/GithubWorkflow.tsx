import React from 'react';

import HeadingBlock from 'app/components/HeadingBlock';
import ExternalLink from 'app/components/ExternalLink';

function GithubWorkflow() {
  return (
    <HeadingBlock
      level="3"
      id="Github-Workflow"
      heading="Github Workflow"
    >
      <b>Tamsui</b>
      {' contains a '}
      <ExternalLink href="https://docs.github.com/en/actions/using-workflows/about-workflows">
        Github workflow
      </ExternalLink>
      {' configured to '}
      <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/.github/workflows/lint-test.yml">
        run the linter and test runner
      </ExternalLink>
      {' on '}
      <ExternalLink href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push">
        push events
      </ExternalLink>
      {'. If this is not desired, '}
      <ExternalLink href="https://docs.github.com/en/actions/using-workflows/disabling-and-enabling-a-workflow#disabling-a-workflow">
        disable the workflow in Github
      </ExternalLink>
      .
    </HeadingBlock>
  );
}

export default GithubWorkflow;
