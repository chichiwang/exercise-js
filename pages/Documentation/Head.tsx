/* global PROJECT_URL */
import React from 'react';

function Head() {
  const url = `${PROJECT_URL}/documentation`;

  return (
    <>
      <meta property="og:type" content="article" />
      <meta property="og:author" content="Chi-chi Wang" />
      <meta property="og:title" content="Developer Documentation for Tamsui: React/Express universal JavaScript boilerplate" />
      <meta property="og:description" content="Set up a universal JavaScript application on the Tamsui boilerplate." />
      <meta property="og:url" content={url} />
      <link href={url} rel="canonical" />
    </>
  );
}

export default Head;
