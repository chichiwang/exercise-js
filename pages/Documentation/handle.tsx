import type { PageHandle } from 'app/types';

const handle: PageHandle = {
  head: {
    title: 'Developer Documentation',
  },
};

export default handle;
