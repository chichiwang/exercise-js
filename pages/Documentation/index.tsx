import React from 'react';
import classNames from 'classnames';

import ContentBlock from 'app/components/ContentBlock';
import HeadingBlock from 'app/components/HeadingBlock';
import ExternalLink from 'app/components/ExternalLink';
import InternalLink from 'app/components/InternalLink';

import StartingAProject from './StartingAProject';
import NPMScripts from './NPMScripts';
import RunningTheApplication from './RunningTheApplication';
import ProjectDirectories from './ProjectDirectories';
import AddingAnApplicationDirectory from './AddingAnApplicationDirectory';
import StaticFiles from './StaticFiles';
import ErrorBoundary from './ErrorBoundary';
import Testing from './Testing';
import ApplicationLayouts from './ApplicationLayouts';
import PullRequestTemplate from './PullRequestTemplate';
import GithubWorkflow from './GithubWorkflow';

import styles from './styles.module.scss';

function Documentation() {
  return (
    <ContentBlock>
      <HeadingBlock
        level="1"
        heading="Developer Documentation"
      >
        <p>
          {'Documentation is also provided in the '}
          <ExternalLink href="https://github.com/chichiwang/tamsui">Github repository</ExternalLink>
          {' in the README file '}
          <ExternalLink href="https://github.com/chichiwang/tamsui/blob/main/docs/README.md">/docs/README.md</ExternalLink>
          .
        </p>
      </HeadingBlock>

      <HeadingBlock
        level="2"
        heading="Contents"
      >
        <ul className={classNames(styles.list, styles.content)}>
          <li>
            <InternalLink to="#Starting-a-Project">Starting a Project</InternalLink>
          </li>
          <li>
            <InternalLink to="#NPM-Scripts">NPM Scripts</InternalLink>
          </li>
          <li>
            <InternalLink to="#Developing-Locally">Developing Locally</InternalLink>
          </li>
          <li>
            <InternalLink to="#Building-for-Production">Building for Production</InternalLink>
          </li>
        </ul>
      </HeadingBlock>

      <StartingAProject />
      <NPMScripts />

      <HeadingBlock
        level="2"
        id="Developing-Locally"
        heading="Developing Locally"
      >
        <ul className={classNames(styles.list, styles.content)}>
          <li>
            <InternalLink to="#Running-the-Application">Running the Application</InternalLink>
          </li>
          <li>
            <InternalLink to="#Project-Directories">Project Directories</InternalLink>
          </li>
          <li>
            <InternalLink to="#Adding-an-Application-Directory">Adding an Application Directory</InternalLink>
          </li>
          <li>
            <InternalLink to="#Static-Files">Static Files</InternalLink>
          </li>
          <li>
            <InternalLink to="#Error-Boundary">Error Boundary</InternalLink>
          </li>
          <li>
            <InternalLink to="#Testing">Testing</InternalLink>
          </li>
          <li>
            <InternalLink to="#Application-Layouts">Application Layouts</InternalLink>
          </li>
          <li>
            <InternalLink to="#Pull-Request-Template">Pull Request Template</InternalLink>
          </li>
          <li>
            <InternalLink to="#Github-Workflow">Github Workflow</InternalLink>
          </li>
        </ul>
      </HeadingBlock>

      <RunningTheApplication />
      <ProjectDirectories />
      <AddingAnApplicationDirectory />
      <StaticFiles />
      <ErrorBoundary />
      <Testing />
      <ApplicationLayouts />
      <PullRequestTemplate />
      <GithubWorkflow />

      <HeadingBlock
        level="2"
        id="Building-for-Production"
        heading="Building for Production"
      >
        <p>
          {'To build for production run '}
          <span className={styles.highlight}>npm build:prod</span>
          {'. The production assets to be deployed will be generated in the '}
          <span className={styles.highlight}>dist/</span>
          {' directory.'}
        </p>

        <p>
          {`To build the production assets and run the Express server in a single
           command, run `}
          <span className={styles.highlight}>npm start</span>
          .
        </p>
      </HeadingBlock>
    </ContentBlock>
  );
}

export default Documentation;
