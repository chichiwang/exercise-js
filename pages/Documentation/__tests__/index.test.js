import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Documentation from '../index';

function RouterWrappedDocumentation() {
  return (
    <MemoryRouter>
      <Documentation />
    </MemoryRouter>
  );
}

describe('Documentation Page Component', () => {
  test('matches snapshot', () => {
    const tree = renderer.create(<RouterWrappedDocumentation />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('contains an h1 tag with the page name', () => {
    render(<RouterWrappedDocumentation />);

    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('Developer Documentation');
  });
});
