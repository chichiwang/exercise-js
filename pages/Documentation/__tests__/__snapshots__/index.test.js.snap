// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documentation Page Component matches snapshot 1`] = `
<div
  className="contentBlock"
>
  <section
    className="headingBlock"
  >
    <div
      className="container"
    >
      <div
        className="wrapper"
      >
        <h1
          className="tag"
        >
          Developer Documentation
        </h1>
      </div>
    </div>
    <article>
      <p>
        Documentation is also provided in the 
        <a
          href="https://github.com/chichiwang/tamsui"
          rel="noreferrer"
          target="_blank"
        >
          Github repository
        </a>
         in the directory 
        <a
          href="https://github.com/chichiwang/tamsui/blob/main/docs/README.md"
          rel="noreferrer"
          target="_blank"
        >
          /docs/README.md
        </a>
        .
      </p>
    </article>
  </section>
  <section
    className="headingBlock"
  >
    <div
      className="container"
    >
      <div
        className="wrapper"
      >
        <h2
          className="tag"
        >
          Contents
        </h2>
      </div>
    </div>
    <article>
      <a
        href="#Starting-a-Project"
      >
        Starting a Project
      </a>
    </article>
  </section>
  <section
    className="headingBlock"
  >
    <div
      className="container"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="wrapper"
      >
        <h2
          className="tag"
          id="Starting-a-Project"
        >
          Starting a Project
        </h2>
        <a
          aria-labelledby="Starting-a-Project"
          className="link"
          href="#Starting-a-Project"
        >
          <svg
            height="1.75rem"
            viewBox="0 0 24 24"
            width="1.75rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2.5}
            >
              <path
                d="M14 11.998C14 9.506 11.683 7 8.857 7H7.143C4.303 7 2 9.238 2 11.998c0 2.378 1.71 4.368 4 4.873a5.3 5.3 0 0 0 1.143.124"
              />
              <path
                d="M10 11.998c0 2.491 2.317 4.997 5.143 4.997h1.714c2.84 0 5.143-2.237 5.143-4.997c0-2.379-1.71-4.37-4-4.874A5.3 5.3 0 0 0 16.857 7"
              />
            </g>
          </svg>
        </a>
      </div>
    </div>
    <article>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Clone the 
          <a
            href="https://github.com/chichiwang/tamsui"
            rel="noreferrer"
            target="_blank"
          >
            repository
          </a>
          .
        </span>
      </label>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Rename the project directory.
        </span>
      </label>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Update the 
          <span
            className="highlight"
          >
            package.json
          </span>
           file.
        </span>
      </label>
      <label
        className="checklistItem indent checklist"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Update the 
          <span
            className="highlight"
          >
            name
          </span>
           field.
        </span>
      </label>
      <label
        className="checklistItem indent checklist"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Update the 
          <span
            className="highlight"
          >
            repository/url
          </span>
           field.
        </span>
      </label>
      <label
        className="checklistItem indent checklist"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Update the 
          <span
            className="highlight"
          >
            bugs/url
          </span>
           field.
        </span>
      </label>
      <label
        className="checklistItem indent checklist"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Update the 
          <span
            className="highlight"
          >
            homepage
          </span>
           field.
        </span>
      </label>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          <a
            href="https://docs.github.com/en/get-started/getting-started-with-git/managing-remote-repositories#changing-a-remote-repositorys-url"
            rel="noreferrer"
            target="_blank"
          >
            Reset the git remote
          </a>
           to your new project repository.
        </span>
      </label>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          If using 
          <a
            href="https://github.com/nvm-sh/nvm"
            rel="noreferrer"
            target="_blank"
          >
            nvm
          </a>
          , run 
          <span
            className="highlight"
          >
            nvm use
          </span>
          .
        </span>
      </label>
      <label
        className="checklistItem indent checklist"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Otherwise, note the contents of 
          <span
            className="highlight"
          >
            .nvmrc
          </span>
           to see the version of 
          <a
            href="https://nodejs.org/"
            rel="noreferrer"
            target="_blank"
          >
            Node.js
          </a>
           this project is built on.
        </span>
      </label>
      <label
        className="checklistItem"
      >
        <input
          checked={false}
          className="checkboxInput"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="checkboxContainer"
        >
          <div
            className="checkbox"
          >
            <div
              className="top"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
            <div
              className="bottom"
            >
              <div
                className="back"
              />
              <div
                className="front"
              />
            </div>
          </div>
        </div>
        <span
          className="labelContent"
        >
          Run 
          <span
            className="highlight"
          >
            npm install
          </span>
          .
        </span>
      </label>
    </article>
  </section>
</div>
`;
